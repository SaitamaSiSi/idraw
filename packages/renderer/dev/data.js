const data = {
  bgColor: "#ffffff",
  elements: [
    // {
    //   name: "rect-001",
    //   x: 5,
    //   y: 5,
    //   w: 100,
    //   h: 50,
    //   type: "rect",
    //   desc: {
    //     bgColor: "#ffeb3b",
    //     borderRadius: 2,
    //     borderWidth: 0,
    //     borderColor: "#ffc107",
    //   },
    // },
    // {
    //   name: "text-002",
    //   x: 40,
    //   y: 40,
    //   w: 100,
    //   h: 60,
    //   // angle: 30,
    //   type: "text",
    //   desc: {
    //     fontSize: 16,
    //     text: [0, 1, 2, 3, 4].map(i => `Hello Text ${i}`).join('\r\n'),
    //     // text: [0, 1, 2, 3, 4].map(i => `Hello Text ${i}`).join(''),
    //     fontWeight: 'bold',
    //     color: "#666666",
    //     borderRadius: 30,
    //     borderWidth: 2,
    //     borderColor: "#ff5722",
    //   },
    // },
    // {
    //   name: "image-003",
    //   x: 80,
    //   y: 80,
    //   w: 160,
    //   h: 80,
    //   type: "image",
    //   desc: {
    //     src: './images/computer.png',
    //   },
    // },
    // {
    //   name: "svg-004",
    //   x: 200 - 5,
    //   y: 150 - 50,
    //   w: 100,
    //   h: 100,
    //   type: "svg",
    //   desc: {
    //     svg: '<svg t="1622524892065" class="icon" viewBox="0 0 1024 1024" version="1.1" xmlns="http://www.w3.org/2000/svg" p-id="9337" width="200" height="200"><path d="M511.6 76.3C264.3 76.2 64 276.4 64 523.5 64 718.9 189.3 885 363.8 946c23.5 5.9 19.9-10.8 19.9-22.2v-77.5c-135.7 15.9-141.2-73.9-150.3-88.9C215 726 171.5 718 184.5 703c30.9-15.9 62.4 4 98.9 57.9 26.4 39.1 77.9 32.5 104 26 5.7-23.5 17.9-44.5 34.7-60.8-140.6-25.2-199.2-111-199.2-213 0-49.5 16.3-95 48.3-131.7-20.4-60.5 1.9-112.3 4.9-120 58.1-5.2 118.5 41.6 123.2 45.3 33-8.9 70.7-13.6 112.9-13.6 42.4 0 80.2 4.9 113.5 13.9 11.3-8.6 67.3-48.8 121.3-43.9 2.9 7.7 24.7 58.3 5.5 118 32.4 36.8 48.9 82.7 48.9 132.3 0 102.2-59 188.1-200 212.9 23.5 23.2 38.1 55.4 38.1 91v112.5c0.8 9 0 17.9 15 17.9 177.1-59.7 304.6-227 304.6-424.1 0-247.2-200.4-447.3-447.5-447.3z" p-id="9338"></path></svg>',
    //   },
    // },

    {
      name: "html-001",
      x: 20,
      y: 40,
      w: 150,
      h: 100,
      type: "html",
      angle: 0,
      desc: {
        html: `
          <style>
          .btn-box {
            margin: 10px 0;
          }
          .btn {
            line-height: 1.5715;
            position: relative;
            display: inline-block;
            font-weight: 400;
            white-space: nowrap;
            text-align: center;
            background-image: none;
            border: 1px solid transparent;
            box-shadow: 0 2px #00000004;
            cursor: pointer;
            user-select: none;
            height: 32px;
            padding: 4px 15px;
            font-size: 14px;
            border-radius: 2px;
            color: #000000d9;
            background: #fff;
            border-color: #d9d9d9;
          }
          .btn-primary {
            color: #fff;
            background: #1890ff;
            border-color: #1890ff;
            text-shadow: 0 -1px 0 rgb(0 0 0 / 12%);
            box-shadow: 0 2px #0000000b;
          }
          </style>
          <div>
            <div class="btn-box">
              <button class="btn">
                <span> Hello &nbsp; Button</span>
              </button>
            </div>
            <div class="btn-box">
              <button class="btn btn-primary">
                <span>Button Primary</span>
              </button> 
            </div>
          </div>
        `,
      },
    },
  ],
};

export function getData() {
  return data;
}
