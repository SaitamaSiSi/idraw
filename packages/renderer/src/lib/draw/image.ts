import { TypeContext, TypeElement, } from 'idraw_zyh_types';
// tempTest
// import { TypeContext, TypeElement, } from '../../../../types/src/index';
import { rotateElement } from '../transform';
import Loader from '../loader';
 

export function drawImage (
  ctx: TypeContext,
  elem: TypeElement<'image'>,
  loader: Loader,
) {
  // const desc = elem.desc as TypeElemDesc['rect'];
  const content = loader.getContent(elem.uuid);
  rotateElement(ctx, elem, () => {
    // ctx.setFillStyle(desc.color);
    // ctx.fillRect(elem.x, elem.y, elem.w, elem.h);
    if (content) {
      // ctx.drawImage(content, 0, 0, elem.w, elem.h, elem.x, elem.y, elem.w, elem.h);
      ctx.drawImage(content, elem.x, elem.y, elem.w, elem.h);
    }
  });
}



// import {
//   TypeContext, 
//   TypeElement,
//   TypeHelperConfig,
//   TypeElemDesc,
// } from '@idraw/types';
// import Loader from '../loader';
// import { drawBox } from './base';

// export function drawImage(
//   ctx: TypeContext,
//   elem: TypeElement<'image'>,
//   loader: Loader,
//   helperConfig: TypeHelperConfig
// ) {
//   const content = loader.getPattern(elem, {
//     forceUpdate: helperConfig?.selectedElementWrapper?.uuid === elem.uuid
//   });
//   drawBox(ctx, elem, content);
// }




 

