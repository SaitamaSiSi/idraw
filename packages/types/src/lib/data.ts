import { TypeElemDesc, TypeElement, TypeElementBase } from './element';

type TypeDataBase = {
  elements: TypeElementBase<keyof TypeElemDesc>[];
  bgColor?: string;
}

type TypeData = {
  elements: TypeElement<keyof TypeElemDesc>[];
  bgColor?: string;
}


export {
  TypeData,
  TypeDataBase,
};