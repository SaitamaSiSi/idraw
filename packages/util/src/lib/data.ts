
export function deepClone(target: any): any {
  function _clone(t: any) {
    const type = is(t);
    if (['Null', 'Number', 'String', 'Boolean', 'Undefined'].indexOf(type) >= 0) {
      return t;
    } else if (type === 'Array') {
      const arr: any[] = [];
      t.forEach((item: any) => {
        arr.push(_clone(item));
      });
      return arr;
    } else if (type === 'Object') {
      const obj: {[key: string]: any} = {};
      const keys = Object.keys(t);
      keys.forEach((key) => {
        obj[key] = _clone(t[key]);
      });
      return obj;
    }
  }
  return _clone(target);
}

function is(data: any): string {
  return Object.prototype.toString.call(data).replace(/[\]|\[]{1,1}/ig, '').split(' ')[1];
}