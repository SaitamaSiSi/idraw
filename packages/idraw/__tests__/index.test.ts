import { requestAnimationFrameMock } from './../../../__tests__/polyfill/requestanimateframe';
import './../../../__tests__/polyfill/image';

import IDraw from './../src';
import { getData } from './data';

function delay(time: number): Promise<void> {
  return new Promise((resolve) => {
    setTimeout(() => {
      resolve();
    }, time);
  });
}

describe('idraw', () => {
  beforeEach(() => {
    requestAnimationFrameMock.reset();
  });

  test('context', async () => {
    document.body.innerHTML = `
      <div id="mount"></div>
    `;
    const opts = {
      width: 600,
      height: 400,
      contextWidth: 600,
      contextHeight: 400,
      devicePixelRatio: 4
    };
    const mount = document.querySelector('#mount') as HTMLDivElement;
    const idraw = new IDraw(mount, opts);
    const data = getData();
    idraw.setData(data, { triggerChangeEvent: true });

    requestAnimationFrameMock.triggerNextAnimationFrame();

    const originCtx = idraw.$getOriginContext2D();
    // @ts-ignore;
    const originCalls = originCtx.__getDrawCalls();
    expect(originCalls).toMatchSnapshot();

    const displayCtx = idraw.$getDisplayContext2D();
    // @ts-ignore;
    const displayCalls = displayCtx.__getDrawCalls();
    expect(displayCalls).toMatchSnapshot();
  });

  test('undo/redo', async () => {
    document.body.innerHTML = `
      <div id="mount"></div>
    `;
    const opts = {
      width: 600,
      height: 400,
      contextWidth: 600,
      contextHeight: 400,
      devicePixelRatio: 4
    };
    const mount = document.querySelector('#mount') as HTMLDivElement;
    const idraw = new IDraw(mount, opts);
    const data = getData();
    idraw.setData(data, { triggerChangeEvent: true });
    idraw.moveDownElement('svg-004');
    idraw.moveDownElement('image-003');
    await delay(10);

    const undo1 = idraw.undo();
    expect(undo1.doRecordCount).toBe(2);
    expect(undo1.data?.elements?.length).toBe(4);

    const undo2 = idraw.undo();
    expect(undo2.doRecordCount).toBe(1);
    expect(undo2.data?.elements?.length).toBe(4);

    const redo1 = idraw.redo();
    expect(redo1.undoRecordCount).toBe(1);
    expect(redo1.data?.elements?.length).toBe(4);

    idraw.moveDownElement('image-003');

    const redo2 = idraw.redo();
    expect(redo2.undoRecordCount).toBe(0);
    expect(redo2.data).toBe(null);

    requestAnimationFrameMock.triggerNextAnimationFrame();

    const originCtx = idraw.$getOriginContext2D();
    // @ts-ignore;
    const originCalls = originCtx.__getDrawCalls();
    expect(originCalls).toMatchSnapshot();

    const displayCtx = idraw.$getDisplayContext2D();
    // @ts-ignore;
    const displayCalls = displayCtx.__getDrawCalls();
    expect(displayCalls).toMatchSnapshot();
  });
});
