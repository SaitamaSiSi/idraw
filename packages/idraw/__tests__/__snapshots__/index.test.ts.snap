// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`idraw_zyh context 1`] = `
Array [
  Object {
    "props": Object {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 120,
            "y": 20,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "cpx1": 860,
            "cpx2": 860,
            "cpy1": 20,
            "cpy2": 460,
            "radius": 100,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 860,
            "cpx2": 20,
            "cpy1": 460,
            "cpy2": 460,
            "radius": 100,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 20,
            "cpx2": 20,
            "cpy1": 460,
            "cpy2": 20,
            "radius": 100,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 20,
            "cpx2": 860,
            "cpy1": 20,
            "cpy2": 20,
            "radius": 100,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "closePath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 120,
            "y": 40,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "cpx1": 840,
            "cpx2": 840,
            "cpy1": 40,
            "cpy2": 440,
            "radius": 80,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 840,
            "cpx2": 40,
            "cpy1": 440,
            "cpy2": 440,
            "radius": 80,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 40,
            "cpx2": 40,
            "cpy1": 440,
            "cpy2": 40,
            "radius": 80,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 40,
            "cpx2": 840,
            "cpy1": 40,
            "cpy2": 40,
            "radius": 80,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "closePath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 560,
            "y": 316,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "cpx1": 1124,
            "cpx2": 1124,
            "cpy1": 316,
            "cpy2": 804,
            "radius": 244,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 1124,
            "cpx2": 316,
            "cpy1": 804,
            "cpy2": 804,
            "radius": 244,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 316,
            "cpx2": 316,
            "cpy1": 804,
            "cpy2": 316,
            "radius": 244,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 316,
            "cpx2": 1124,
            "cpy1": 316,
            "cpy2": 316,
            "radius": 244,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "closePath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 560,
            "y": 320,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "cpx1": 1120,
            "cpx2": 1120,
            "cpy1": 320,
            "cpy2": 800,
            "radius": 240,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 1120,
            "cpx2": 320,
            "cpy1": 800,
            "cpy2": 800,
            "radius": 240,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 320,
            "cpx2": 320,
            "cpy1": 800,
            "cpy2": 320,
            "radius": 240,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 320,
            "cpx2": 1120,
            "cpy1": 320,
            "cpy2": 320,
            "radius": 240,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "closePath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "Hello Text",
      "x": 715,
      "y": 520,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;

exports[`idraw_zyh context 2`] = `
Array [
  Object {
    "props": Object {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "dHeight": 1600,
      "dWidth": 2400,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="1600"
        width="2400"
      />,
      "sHeight": 1600,
      "sWidth": 2400,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 1600,
      "dWidth": 2400,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="1600"
        width="2400"
      />,
      "sHeight": 1600,
      "sWidth": 2400,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;

exports[`idraw_zyh undo/redo 1`] = `
Array [
  Object {
    "props": Object {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 120,
            "y": 20,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "cpx1": 860,
            "cpx2": 860,
            "cpy1": 20,
            "cpy2": 460,
            "radius": 100,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 860,
            "cpx2": 20,
            "cpy1": 460,
            "cpy2": 460,
            "radius": 100,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 20,
            "cpx2": 20,
            "cpy1": 460,
            "cpy2": 20,
            "radius": 100,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 20,
            "cpx2": 860,
            "cpy1": 20,
            "cpy2": 20,
            "radius": 100,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "closePath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 120,
            "y": 40,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "cpx1": 840,
            "cpx2": 840,
            "cpy1": 40,
            "cpy2": 440,
            "radius": 80,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 840,
            "cpx2": 40,
            "cpy1": 440,
            "cpy2": 440,
            "radius": 80,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 40,
            "cpx2": 40,
            "cpy1": 440,
            "cpy2": 40,
            "radius": 80,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 40,
            "cpx2": 840,
            "cpy1": 40,
            "cpy2": 40,
            "radius": 80,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "closePath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 560,
            "y": 316,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "cpx1": 1124,
            "cpx2": 1124,
            "cpy1": 316,
            "cpy2": 804,
            "radius": 244,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 1124,
            "cpx2": 316,
            "cpy1": 804,
            "cpy2": 804,
            "radius": 244,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 316,
            "cpx2": 316,
            "cpy1": 804,
            "cpy2": 316,
            "radius": 244,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 316,
            "cpx2": 1124,
            "cpy1": 316,
            "cpy2": 316,
            "radius": 244,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "closePath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "stroke",
  },
  Object {
    "props": Object {
      "fillRule": "nonzero",
      "path": Array [
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "beginPath",
        },
        Object {
          "props": Object {
            "x": 560,
            "y": 320,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "moveTo",
        },
        Object {
          "props": Object {
            "cpx1": 1120,
            "cpx2": 1120,
            "cpy1": 320,
            "cpy2": 800,
            "radius": 240,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 1120,
            "cpx2": 320,
            "cpy1": 800,
            "cpy2": 800,
            "radius": 240,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 320,
            "cpx2": 320,
            "cpy1": 800,
            "cpy2": 320,
            "radius": 240,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {
            "cpx1": 320,
            "cpx2": 1120,
            "cpy1": 320,
            "cpy2": 320,
            "radius": 240,
          },
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "arcTo",
        },
        Object {
          "props": Object {},
          "transform": Array [
            1,
            0,
            0,
            1,
            0,
            0,
          ],
          "type": "closePath",
        },
      ],
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fill",
  },
  Object {
    "props": Object {
      "maxWidth": null,
      "text": "Hello Text",
      "x": 715,
      "y": 520,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "fillText",
  },
]
`;

exports[`idraw_zyh undo/redo 2`] = `
Array [
  Object {
    "props": Object {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "height": 1600,
      "width": 2400,
      "x": 0,
      "y": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "clearRect",
  },
  Object {
    "props": Object {
      "dHeight": 1600,
      "dWidth": 2400,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="1600"
        width="2400"
      />,
      "sHeight": 1600,
      "sWidth": 2400,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
  Object {
    "props": Object {
      "dHeight": 1600,
      "dWidth": 2400,
      "dx": 0,
      "dy": 0,
      "img": <canvas
        height="1600"
        width="2400"
      />,
      "sHeight": 1600,
      "sWidth": 2400,
      "sx": 0,
      "sy": 0,
    },
    "transform": Array [
      1,
      0,
      0,
      1,
      0,
      0,
    ],
    "type": "drawImage",
  },
]
`;
